cni: Follow CNI spec by using `(containerID, ifName)` as unique endpoint identifier

gandro opened the pull request:
The [CNI spec states](https://www.cni.dev/docs/spec/#section-3-execution-of-network-configurations) that a attachment is uniquely identified by the tuple consisting of both the container runtime container ID and the container interface name. Since a Cilium endpoint corresponds to a network attachment, the  container ID alone is not sufficient to uniquely identify an endpoint.

Therefore, this commit deprecates the concept of the container ID as a unique identifier for an endpoint, and instead introduces the "CNI attachment ID", which consists of both the containerID *and* container interface name, which is guaranteed to be unique according to the CNI spec.

Consequently, this PR also removes the restriction that there is a 1:1 mapping between a pod and an endpoint. To achieve this, we remove the `LookupPodName` helper in the agent and replace it with a `LookupCEPName` (for correlating with a CEP) and `GetEndpointsByPodName` (for correlating with a pod) helpers instead.

This change should be fully backwards compatible for endpoints which have a single interface.

 - Review per commit.
 - Co-authored-by @tklauser

tklauser committed:
api: clarify Interface{Index,Name} field comments

Document that these are the interface index and name in the host network
namespace.

Signed-off-by: Tobias Klauser <tobias@isovalent.com>
Signed-off-by: Sebastian Wicki <sebastian@isovalent.com>

tklauser committed:
api: add endpoint fields for container interface

Add a field to the EndpointChangeRequest and EndpointNetworking types
which is used to store the name of the network device in the container
network namespace. This will be used in a subsequent commit to allow
lookup via CNI attachment id (which contains the container interface
name).

Signed-off-by: Tobias Klauser <tobias@isovalent.com>
Signed-off-by: Sebastian Wicki <sebastian@isovalent.com>

tklauser committed:
endpoint: populate container interface fields

Populate the endpoints container interface fields from the respective
fields set in the Cilium API models types. We also add the field to the
serialized endpoint used for endpoint restoration.

They will be used in a subsequent commit.

Signed-off-by: Tobias Klauser <tobias@isovalent.com>
Signed-off-by: Sebastian Wicki <sebastian@isovalent.com>

tklauser committed:
operator/cmd: use OwnerReference name and namespace for Pod lookup

Currently, we use the CEP namespace and name to look up the owning Pod
for a CEP. Switch the lookup to use the namespace and name. Currently,
these are the same values but it makes it semantically cleaner to use
the OwnerReference.

Signed-off-by: Tobias Klauser <tobias@isovalent.com>
Signed-off-by: Sebastian Wicki <sebastian@isovalent.com>

gandro committed:
plugins/cilium-cni: set container network interface name

Co-authored-by: Tobias Klauser <tobias@isovalent.com>
Signed-off-by: Sebastian Wicki <sebastian@isovalent.com>

gandro committed:
endpoint: Allow multiple endpoints per pod

This commit removes the restriction that there is a 1:1 mapping between
a pod and an endpoint. To achieve this, we remove the `LookupPodName`
helper in the agent and replace it with a `LookupCEPName` (for
correlating with a CEP) and `GetEndpointsByPodName` (for correlating
with a pod) helpers instead.

Lookups on the pod name via API continue to be supported for backwards
compatibility, if the endpoint has not opted out of legacy identifiers.
The API also allows lookup by `cep-name` with this change, but similar
to other mostly internal identifiers (such as the `cilium-global`
prefix) we do not announce this ID in the `EndpointIdentifiers` model.

Co-authored-by: Tobias Klauser <tobias@isovalent.com>
Signed-off-by: Tobias Klauser <tobias@isovalent.com>
Signed-off-by: Sebastian Wicki <sebastian@isovalent.com>

gandro committed:
api: Add `cni-attachment-id` as endpoint identifier

This commit adds a new endpoint identifier called `cni-attachment-id`.
It it intended to replace the container ID and container name as a
unique identifier, because the CNI spec allows the runtime to invoke a
CNI ADD/DEL for the same container ID twice. Therefore, instead of the
container ID, we now recommend using the CNI attachment ID, which
consists of the `(containerID, ifName)` pair.

The container ID and container name IDs will continue to work for
primary attachments, but are now deprecated.

This commit only changes the API definition. The new ID will be hooked
up in a subsequent commit.

Signed-off-by: Sebastian Wicki <sebastian@isovalent.com>

gandro committed:
cni: Follow CNI spec by using `(containerID, ifName)` as endpoint ID

The CNI spec states [1] that a attachment is uniquely identified by the
tuple consisting of both the container runtime container ID and the
container interface name. Since a Cilium endpoint corresponds to a
network attachment, the container ID alone is not sufficient to uniquely
identify an endpoint.

Therefore, this commit deprecates the concept of the container ID as a
unique identifier for an endpoint, and instead introduces the "CNI
attachment ID", which consists of both the containerID and container
interface name. This is guaranteed to be unique according to the CNI
spec.

The container id (and container name) continue to be supported in the
Cilium API when an endpoint did not opt out of those legacy identifiers,
in which case it is also implicitly assumed that there is only one
endpoint per container.

[1] https://www.cni.dev/docs/spec/#section-3-execution-of-network-configurations

Signed-off-by: Sebastian Wicki <sebastian@isovalent.com>

gandro committed:
api: Deprecate `pod-name` as endpoint id

This commit deprecates the `pod-name` lookup in the Cilium API and
replaces it with a `cep-name` lookup instead. As with the previous
change in this commit series, the `pod-name` may no longer uniquely
identify an endpoint if a pod has multiple network attachments.

`pod-name` will continue to work as an endpoint id for the endpoint of
the primary network interface.

This commit only changes the API definition. The new ID will be hooked
up in a subsequent commit.

Co-authored-by: Tobias Klauser <tobias@isovalent.com>
Signed-off-by: Tobias Klauser <tobias@isovalent.com>
Signed-off-by: Sebastian Wicki <sebastian@isovalent.com>

asauber reviewed (approved):
CLI-owned changes LGTM. I haven't reviewed for correctness of the new endpoint logic.

asauber commented on api/v1/openapi.yaml:
```diff
@@ -1796,23 +1804,32 @@ definitions:
+      container-interface-name:
+        description: Name of network device in container netns
+        type: string
+      container-interface-is-secondary:
```
Naming this "not primary" rather than "secondary", seems to make more sense. Not a blocker for me.

tommyp1ckles commented on api/v1/openapi.yaml:
```diff
@@ -1820,11 +1821,14 @@ definitions:
+        type: string
       container-id:
-        description: ID assigned by container runtime
+        description: ID assigned by container runtime (deprecated, may not be unique)
```
After we go up to version v1.16 could we remove this altogether?

gandro replied:
I don't think we can remove this until we remove the docker plugin (ref https://github.com/cilium/cilium/issues/27022)

tommyp1ckles commented on pkg/endpoint/identifiers.go:
```diff
@@ -41,6 +41,21 @@ func (e *Endpoint) GetK8sNamespaceAndPodName() string {
+	if e.containerIfName != "" {
+		return e.containerID + ":" + e.containerIfName
+	}
+	return e.containerID
```
should we ensure that this is a primary iface endpoint in this case?

gandro replied:
Yes good point, I'll look into it.

gandro replied:
Fixed by removing the `secondary interface` flag instead

tommyp1ckles commented on pkg/endpoint/identifiers.go:
```diff
@@ -41,6 +41,23 @@ func (e *Endpoint) GetK8sNamespaceAndPodName() string {
 
+// GetK8sCEPResourceName returns the corresponding K8s CiliumEndpoint resource name
+// for this endpoint (without the namespace)
+func (e *Endpoint) GetK8sCEPResourceName() string {
```
Can we document that its expected that this must return an empty string if the endpoint hasn't synced its PodName yet?

gandro replied:
Not sure I follow. The pod name is set upon initial creation of the endpoint, so this should never be empty if the corresponding endpoint is created for a pod In CNI ADD.

It will always be empty for endpoints without a pod (e.g. host endpoint, health endpoint), and it will always be populated for endpoints backed by a pod, _even_ if we haven't synced the corresponding pod yet.

See also https://github.com/cilium/cilium/commit/e20b16d70843530aefad9ac8c3ae0834c7b6b057

gandro replied:
Documented that it can return an empty string!

tommyp1ckles commented on pkg/endpoint/api.go:
```diff
@@ -145,17 +145,16 @@ func (e *Endpoint) getModelEndpointIdentitiersRLocked() *models.EndpointIdentifi
 		identifiers.ContainerName = e.containerName
+		identifiers.PodName = e.GetK8sNamespaceAndPodName()
+		identifiers.K8sPodName = e.K8sPodName
+		identifiers.K8sNamespace = e.K8sNamespace
```
Not really related specifically to this but this struct model is kind of strange, I expected everything in here to uniquely identify the endpoint but namespace does not do that on its own.

Is this even used anymore?

gandro replied:
Yeah, this is a bit odd. The K8sNamespace is indeed not unique (and in fact we don't have a way to just pass a namespace into the `Lookup` method). I think this exists so K8s pod and K8s namespace are kept separate in the EndpointIndentifiers struct (they replaced the deprecated PodName field).

But I'm not sure they are actually used, I see no usages in our code. But they are part of the CEP CRD, so I didn't want to change them.

gandro replied:
Unless you feel otherwise, I'll leave this as is for now.

tommyp1ckles commented on pkg/endpoint/identifiers.go:
```diff
@@ -41,6 +41,23 @@ func (e *Endpoint) GetK8sNamespaceAndPodName() string {
 
+// GetK8sCEPResourceName returns the corresponding K8s CiliumEndpoint resource name
+// for this endpoint (without the namespace)
+func (e *Endpoint) GetK8sCEPResourceName() string {
```
small thought: the CEPResource name seems a bit redundant to me

gandro replied:
Fixed by renaming to `GetK8sCEPName`

tommyp1ckles commented on pkg/k8s/watchers/endpointsynchronizer.go:
```diff
@@ -95,11 +95,11 @@ func (epSync *EndpointSynchronizer) RunK8sCiliumEndpointSync(e *endpoint.Endpoin
+				// The CEP name is derived from the K8sPodName and K8sNamespace.
+				// Those are not always available when an endpoint is first created, so we collect them here.
+				cepName := e.GetK8sCEPResourceName()
+				if cepName == "" {
```
The function comment says that K8sPodName is const but this would imply that it gets set later.

gandro replied:
Oh yeah good catch. This comment is false since https://github.com/cilium/cilium/commit/e20b16d70843530aefad9ac8c3ae0834c7b6b057, I'll fix it.

gandro replied:
Fixed by updating the comment to reflect reality

zacharysarah reviewed (approved):
LGTM for docs

gandro committed:
api: Add flag to disable legacy identifiers

This will allow endpoints to opt-in to new endpoint identifier system,
which requires endpoints to be looked up either using the CNI attachment
ID or the CEP name (both of which set the container interface name).
This will allow multiple endpoints within the same pod or container.

If an endpoint creation request does not explicitly opt-in, we will
continue to support addressing it by the container ID or pod name alone
for backwards-compatibility reasons.

Signed-off-by: Sebastian Wicki <sebastian@isovalent.com>

joamaki reviewed (approved):
operator changes LGTM

tommyp1ckles commented on pkg/endpointmanager/manager.go:
```diff
@@ -330,14 +333,27 @@ func (mgr *endpointManager) LookupIP(ip netip.Addr) (ep *endpoint.Endpoint) {
 
+// GetEndpointsByPodName looks up endpoints by namespace + pod name
+func (mgr *endpointManager) GetEndpointsByPodName(namespacedName string) []*endpoint.Endpoint {
+	mgr.mutex.RLock()
```
nit: can use defer here.

tommyp1ckles commented on pkg/endpointmanager/manager.go:
```diff
@@ -330,14 +333,27 @@ func (mgr *endpointManager) LookupIP(ip netip.Addr) (ep *endpoint.Endpoint) {
+func (mgr *endpointManager) GetEndpointsByPodName(namespacedName string) []*endpoint.Endpoint {
+	mgr.mutex.RLock()
+	eps := make([]*endpoint.Endpoint, 0, 1)
+	for _, ep := range mgr.endpoints {
```
this is used by watcher Pod updates I believe, we're now iterating every endpoint upon each (local) update.  Probably is fine - but I wonder if there's cases where this might cause slowdowns.

gandro replied:
That is a valid concern. The current indexing unfortunately is not powerful enough to avoid that. I looked into it, and we would have to introduce another `endpointsByPodName map[string]map[uint16]*endpoint.Endpoint` (or similar) index in the endpoint manager to fix it, which seems a bit intrusive. This could be a great use-case for statedb.

Given that the number of endpoints on a node should be in the order of few hundred pods at most I'm not too worried and will leave it as is for now.

tommyp1ckles commented on pkg/k8s/watchers/endpointsynchronizer.go:
```diff
@@ -95,11 +95,11 @@ func (epSync *EndpointSynchronizer) RunK8sCiliumEndpointSync(e *endpoint.Endpoin
+				// They should always be available if an endpoint belongs to a pod.
+				cepName := e.GetK8sCEPName()
+				if cepName == "" {
+					scopedLog.Debug("Skipping CiliumEndpoint update because it has no k8s cep name")
```
It seems like this case would should now be unexpected, and potentially a bug?  Perhaps we should error log this (and maybe return err to the controller?)

gandro replied:
Good point, I just checked and we are running this controller for the health and host endpoint which do not have a pod backing it. So emitting a warning or error would cause a lot of unnecessary noise.

We should probably fix that (i.e. not start the controller for such endpoints), but I'll leave that up for future work. I've opened an issue: https://github.com/cilium/cilium/issues/27128

tommyp1ckles commented on api/v1/models/endpoint_change_request.go:
```diff
@@ -40,6 +40,11 @@ type EndpointChangeRequest struct {
 
+	// Disables lookup using legacy endpoint identifiers (container name,
+	// container id, pod name) for this endpoint
+	//
```
nit: extra comment line here

gandro replied:
Took me a while to figure this out, since it's auto generated. For future reference, the trick is to use ` >-` in openapi.yaml for the description

tommyp1ckles reviewed (approved):
k8s changes lgtm, just a few minor things.
